src/watcher/store.ts

```diff
@@ -59,23 +69,33 @@ export class AppointmentStore extends StartStopService {
     public addOrUpdateByLocator(appointment: Appointment): Promise<void> {
         // As we are accessing data structures by state locator, we make sure to acquire a lock on it
         return this.stateLocatorLockManager.withLock(appointment.locator, async () => {
+            const batch = this.subDb.batch();
+
             const currentAppointment = this.mAppointmentsByLocator.get(appointment.locator);
-            // is there a current appointment
+
+            // is there an existing appointment
             if (currentAppointment) {
-                if (appointment.nonce > currentAppointment.nonce) this.mAppointmentsById.delete(currentAppointment.id);
+                // make sure that the nonce is larger than the previous one
+                if (appointment.nonce > currentAppointment.nonce) {
+                    this.mAppointmentsById.delete(currentAppointment.id);	
+                    this.mAppointmentsByCustomerAddress.deleteFromSet(appointment.customerAddress, currentAppointment);
+                }
                 else {
-                    throw new ApplicationError(appointment.formatLog(`Nonce ${appointment.nonce} is lower than current appointment ${currentAppointment.locator} nonce ${currentAppointment.nonce}`)) //prettier-ignore
+                    throw new ApplicationError(appointment.formatLog(`Nonce ${appointment.nonce} is not larger than current appointment ${currentAppointment.locator} nonce ${currentAppointment.nonce}`)) //prettier-ignore
                 }
+
+                batch.del(currentAppointment.id);
             }
 
+            batch.put(appointment.id, Appointment.toIAppointment(appointment));
+
             // update the db
-            const batch = this.subDb.batch().put(appointment.id, Appointment.toIAppointment(appointment));
-            if (currentAppointment) await batch.del(currentAppointment.id).write();
-            else await batch.write();
+            await batch.write();
 
-            // add the new appointment
+            // add the new appointment, and replace an old one
             this.mAppointmentsByLocator.set(appointment.locator, appointment);
             this.mAppointmentsById.set(appointment.id, appointment);
+            this.mAppointmentsByCustomerAddress.addToSet(appointment.customerAddress, appointment);
         });
     }
```

Comment:
In previous attempts to write this function, I was splitting it in three blocks:
1) validation (e.g. throw an error if someething is wrong)
2) write the batch to db
3) update the in-memory structures.

The reason is that if writing the batch throws an error, this would make sure that also in-memory changes are not done (while the code now does some in-memory changes before writing the batch).
It's definitely easier to read this version of the code, though (I had to put two conditional blocks over currentAppointment for that reason). Do you think that the in-memory consistency in case of db failure is a desirable property?

Reply:
I agree, updated.